### PR TITLE
DateTime uses strptime during deserialization

### PR DESCRIPTION
**Pull request by bartaelterman**
Note that this fix will solely rely on the DateTime format given. No deserialization attempt is made using dateutil.

### CODE REVIEW

**Comment by sloria** [comment]
It'd be nice to have the option of using dateutil for parsing. Perhaps we should only use strptime if `dateformat` is a string that isn't one of the keys of `DATETIME_DESERIALIZATION_FUNCS`.

**Comment by bartaelterman** [comment]
That is what happens now. The check is still present [here](https://github.com/bartaelterman/marshmallow/blob/dev/marshmallow/fields.py#L766). However, I think what we really want is (in pseudo code):

```
if value is in DATETIME_DESERIALIZATION_FUNCS:
    parse with that function
elif value is a valid strptime time formatting string:
    parse with strptime
else
    parse with dateutil
```

However:
1. From the API perspective, I think that results in weird behaviour. Either the user sets the `format` parameter to `rfc` or `iso` (which will result in parsing with the function in `DATETIME_DESERIALIZATION_FUNCS`), or he gives a time formatting string, which will result in parsing with `strptime`. If the parameter is not set, `iso` is used as default. So from a user perspective, to what value should I set the `format` parameter to have dateutil do the parsing? Some random string not being `iso`, `rfc` or a time formatting string and not `None`? Maybe the best option would then be to override the behaviour in case the parameter is `None`.
2. I'm not sure how to check whether the given string is a valid strptime formatting string anyway.

So I agree that parsing with dateutil could be useful, but we should give it a proper place in the API.

**Comment by sloria** [comment]
I think it makes sense to parse with dateutil if `format=None`. So the pseudocode becomes:

```
if value is in DATETIME_DESERIALIZATION_FUNCS:
    parse with that function
elif value is a string:
    parse with strptime
else
    parse with dateutil
```

The default `format` would be changed to `None`.

**Comment by bartaelterman** [comment]
Ok, I'll change it.

**Comment by sloria** [comment]
@bartaelterman Can you look into the failing tests?

**Comment by bartaelterman** [comment]
Will do so asap.

**Comment by bartaelterman** [comment]
Fixed two tests. Three yet to fix.

File: marshmallow/fields.py [hunk 1]
```diff
@@ -803,17 +805,21 @@ def _deserialize(self, value):
...
-        self.dateformat = self.dateformat or self.DEFAULT_FORMAT
         func = self.DATEFORMAT_DESERIALIZATION_FUNCS.get(self.dateformat)
         if func:
             try:
                 return func(value)
             except (TypeError, AttributeError, ValueError):
                 raise err
+        elif type(value) is str and self.dateformat:
+            try:
+                return dt.datetime(*time.strptime(value, self.dateformat)[0:6])
```
Comment (sloria) [inline marshmallow/fields.py:816]: Why not use `dt.datetime.strptime`?

File: marshmallow/fields.py [hunk 2]
```diff
@@ -803,17 +805,21 @@ def _deserialize(self, value):
...
         if not value:  # Falsy values, e.g. '', None, [] are not valid
             raise err
-        self.dateformat = self.dateformat or self.DEFAULT_FORMAT
         func = self.DATEFORMAT_DESERIALIZATION_FUNCS.get(self.dateformat)
         if func:
             try:
                 return func(value)
             except (TypeError, AttributeError, ValueError):
                 raise err
+        elif type(value) is str and self.dateformat:
```
Comment (sloria) [inline marshmallow/fields.py:814]: I'm not sure that the typechecking is necessary. We can assume that `self.dateformat` is either a string or `None`.

Comment (justanr) [inline marshmallow/fields.py:814]: Especially because of the difference between `""` and `u""` in Python 2.

**Comment by sloria** [comment]
It appears that https://github.com/marshmallow-code/marshmallow/pull/262 attempts to solve the same issue.

@bartaelterman Would that PR solve the issue effectively? Is there anything present in this approach that is missing in #262 ?

**Comment by bartaelterman** [comment]
Nope, that's exactly the same thing.

**Pull request closed without merging**